File: scratch/test_photo_removal.py
import os
from PIL import Image

def get_pixel(pixels, x, y, width):
    return pixels[x, y]

def process_image(img_path, output_path, threshold=55):
    # Open image
    img = Image.open(img_path)
    img = img.convert('RGB')
    
    # Scale down to max 400px (like our JS code)
    max_dim = 400
    w, h = img.size
    if w > max_dim or h > max_dim:
        if w > h:
            h = int((h * max_dim) / w)
            w = max_dim
        else:
            w = int((w * max_dim) / h)
            h = max_dim
        img = img.resize((w, h), Image.Resampling.LANCZOS)
    
    pixels = img.load()
    width, height = img.size
    
    # Sample background color from corners and top-middle
    samples = [
        get_pixel(pixels, 0, 0, width),
        get_pixel(pixels, width - 1, 0, width),
        get_pixel(pixels, 0, min(10, height - 1), width),
        get_pixel(pixels, width - 1, min(10, height - 1), width),
        get_pixel(pixels, int(width / 2), 0, width)
    ]
    
    bg_r = sum(s[0] for s in samples) // len(samples)
    bg_g = sum(s[1] for s in samples) // len(samples)
    bg_b = sum(s[2] for s in samples) // len(samples)
    
    print(f"Detected background color: RGB({bg_r}, {bg_g}, {bg_b})")
    
    # BFS queue for flood fill
    visited = [0] * (width * height)
    queue = []
    
    # Initialize queue with border pixels
    for x in range(width):
        queue.append((x, 0))
        visited[0 * width + x] = 1
    for y in range(1, height):
        queue.append((0, y))
        queue.append((width - 1, y))
        visited[y * width + 0] = 1
        visited[y * width + (width - 1)] = 1
        
    def is_color_similar(r, g, b):
        dr = r - bg_r
        dg = g - bg_g
        db = b - bg_b
        import math
        return math.sqrt(dr*dr + dg*dg + db*db) < threshold

    # Create new output image with RGBA
    out_img = Image.new('RGBA', (width, height), (255, 255, 255, 255))
    out_pixels = out_img.load()
    
    # Copy original to output first
    for y in range(height):
        for x in range(width):
            r, g, b = pixels[x, y]
            out_pixels[x, y] = (r, g, b, 255)
            
    head = 0
    while head < len(queue):
        x, y = queue[head]
        head += 1
        r, g, b = pixels[x, y]
        
        if is_color_similar(r, g, b):
            # Mark as background (turn white)
            out_pixels[x, y] = (255, 255, 255, 255)
            
            # Neighbors
            neighbors = [
                (x - 1, y),
                (x + 1, y),
                (x, y - 1),
                (x, y + 1)
            ]
            for nx, ny in neighbors:
                if 0 <= nx < width and 0 <= ny < height:
                    n_idx = ny * width + nx
                    if not visited[n_idx]:
                        visited[n_idx] = 1
                        queue.append((nx, ny))
                        
    out_img.save(output_path, 'PNG')
    print("Image processed and saved.")

# Run processing
img_dir = r"C:\Users\salim\.gemini\antigravity\brain\7dbd627f-6ee4-40d8-8e11-80bf7ab183a3"
process_image(
    os.path.join(img_dir, "media__1782372962202.jpg"),
    os.path.join(img_dir, "media__1782372962202_processed.png")
)

File: scratch/test_photo_removal.js
const sharp = require('sharp');
const fs = require('fs');

const imgDir = "C:/Users/salim/.gemini/antigravity/brain/7dbd627f-6ee4-40d8-8e11-80bf7ab183a3";
const inputPath = `${imgDir}/media__1782372962202.jpg`;
const outputPath = `${imgDir}/media__1782372962202_processed.png`;

sharp(inputPath)
  .resize(400, 400, { fit: 'inside' })
  .raw()
  .toBuffer({ resolveWithObject: true })
  .then(({ data, info }) => {
     const { width, height, channels } = info;
     const outputData = Buffer.alloc(width * height * 4); // RGBA output
     
     const getPixel = (x, y) => {
         const idx = (y * width + x) * channels;
         return { r: data[idx], g: data[idx+1], b: data[idx+2] };
     };
     
     // Sample 8 background anchor points (corners and upper borders)
     const samples = [
         getPixel(0, 0), // Top-Left
         getPixel(width - 1, 0), // Top-Right
         getPixel(0, Math.min(10, height - 1)), // Upper Left
         getPixel(width - 1, Math.min(10, height - 1)), // Upper Right
         getPixel(Math.round(width / 2), 0), // Top Middle
         getPixel(Math.round(width / 4), 0), // Top Left-Middle
         getPixel(Math.round(3 * width / 4), 0), // Top Right-Middle
         getPixel(0, Math.min(40, height - 1)), // Mid Left
         getPixel(width - 1, Math.min(40, height - 1)) // Mid Right
     ];
     
     console.log("Background anchors sampled:", samples.length);
     
     const visited = new Uint8Array(width * height);
     const queue = [];
     
     // Initialize BFS queue with border pixels (top, left, right borders)
     for (let x = 0; x < width; x++) {
         queue.push({ x, y: 0 });
         visited[0 * width + x] = 1;
     }
     for (let y = 1; y < height; y++) {
         queue.push({ x: 0, y });
         queue.push({ x: width - 1, y });
         visited[y * width] = 1;
         visited[y * width + (width - 1)] = 1;
     }
     
     // Threshold: distance limit to match any background anchor
     const threshold = 55;
     
     const isColorSimilar = (r, g, b) => {
         // Check distance against ALL background anchors
         for (let i = 0; i < samples.length; i++) {
             const s = samples[i];
             const dr = r - s.r;
             const dg = g - s.g;
             const db = b - s.b;
             const dist = Math.sqrt(dr*dr + dg*dg + db*db);
             if (dist < threshold) {
                 return true;
             }
         }
         return false;
     };
     
     // Copy original pixels to output buffer (RGBA)
     for (let y = 0; y < height; y++) {
         for (let x = 0; x < width; x++) {
             const idx = (y * width + x) * channels;
             const outIdx = (y * width + x) * 4;
             outputData[outIdx] = data[idx];
             outputData[outIdx+1] = data[idx+1];
             outputData[outIdx+2] = data[idx+2];
             outputData[outIdx+3] = 255;
         }
     }
     
     let head = 0;
     let cleanedCount = 0;
     while (head < queue.length) {
         const { x, y } = queue[head++];
         const idx = (y * width + x) * channels;
         const outIdx = (y * width + x) * 4;
         const r = data[idx];
         const g = data[idx+1];
         const b = data[idx+2];
         
         if (isColorSimilar(r, g, b)) {
             // Turn to solid white
             outputData[outIdx] = 255;
             outputData[outIdx+1] = 255;
             outputData[outIdx+2] = 255;
             outputData[outIdx+3] = 255;
             cleanedCount++;
             
             // 4-way neighbors
             const neighbors = [
                 { x: x - 1, y },
                 { x: x + 1, y },
                 { x, y: y - 1 },
                 { x, y: y + 1 }
             ];
             
             neighbors.forEach(n => {
                 if (n.x >= 0 && n.x < width && n.y >= 0 && n.y < height) {
                     const nIdx = n.y * width + n.x;
                     if (!visited[nIdx]) {
                         visited[nIdx] = 1;
                         queue.push(n);
                     }
                 }
             });
         }
     }
     
     console.log(`Flood fill completed. Cleaned pixels: ${cleanedCount} / ${width * height}`);
     
     sharp(outputData, { raw: { width, height, channels: 4 } })
       .png()
       .toFile(outputPath)
       .then(() => {
           console.log("Processed image saved successfully to:", outputPath);
       })
       .catch(err => {
           console.error("Error saving output image:", err);
       });
  })
  .catch(err => {
      console.error("Error processing image:", err);
  });
